File: CosmosCore/src/com/cornchipss/cosmos/blocks/BlockFace.java
package com.cornchipss.cosmos.blocks;

import org.joml.Vector3f;

/**
 * The faces of each block
 */
public enum BlockFace
{
	FRONT(0), 
	BACK(1),
	TOP(2),
	BOTTOM(3),
	RIGHT(4),
	LEFT(5),
	MAX(5),
	UNKNOWN(-1);
	
	private int f;
	
	private BlockFace(int f)
	{
		this.f = f;
	}
	
	/**
	 * <ul>
	 * 	<li>0: FRONT</li>
	 * 	<li>1: FRONT</li>
	 * 	<li>2: FRONT</li>
	 * 	<li>3: FRONT</li>
	 * 	<li>4: FRONT</li>
	 * 	<li>5: FRONT</li>
	 * 	<li>-1: UNKNOWN</li>
	 * </ul>
	 */
	public int getValue()
	{
		return f;
	}

	/**
	 * <p>The block face that corresponds to that index</p>
	 * <ul>
	 * 	<li>0: FRONT</li>
	 * 	<li>1: FRONT</li>
	 * 	<li>2: FRONT</li>
	 * 	<li>3: FRONT</li>
	 * 	<li>4: FRONT</li>
	 * 	<li>5: FRONT</li>
	 * 	<li>?: UNKNOWN</li>
	 * </ul>
	 * @param i The index
	 * @return The block face that corresponds to that index
	 */
	public static BlockFace fromFaceIndex(int i)
	{
		switch(i)
		{
		case 0:
			return FRONT;
		case 1:
			return BACK;
		case 2:
			return TOP;
		case 3:
			return BOTTOM;
		case 4:
			return RIGHT;
		case 5:
			return LEFT;
		default:
			return UNKNOWN;
		}
	}
	
	/**
	 * The position relative to the block's center assuming no rotation
	 * @return The position relative to the block's center assuming no rotation
	 */
	public Vector3f getRelativePosition()
	{
		switch(this)
		{
		case FRONT:
			return new Vector3f(0, 0, .5f);
		case BACK:
			return new Vector3f(0, 0, -.5f);
		case TOP:
			return new Vector3f(0, .5f, 0);
		case BOTTOM:
			return new Vector3f(0, -.5f, 0);
		case RIGHT:
			return new Vector3f(.5f, 0, 0);
		case MAX:
		case LEFT:
			return new Vector3f(-.5f, 0, 0);
		default:
			return new Vector3f(0, 0, 0);
		}
	}
}

File: cosmos_core/src/block/block_face.rs
//! The 6 faces of a block. These directly correspond to textures, but do not always imply
//! a certain [`BlockDirection`]due to the possibility of a [`super::block_rotation::BlockRotation`].

use std::fmt::Display;

use bevy::reflect::Reflect;
use serde::{Deserialize, Serialize};

use super::block_direction::BlockDirection;

#[derive(Debug, PartialEq, Eq, Reflect, Default, Copy, Clone, Serialize, Deserialize, Hash)]
/// Represents the different faces of a block.
///
/// Even non-cube blocks will have this.
pub enum BlockFace {
    /// +X
    Right,
    /// -X
    Left,
    /// +Y
    #[default]
    Top,
    /// -Y
    Bottom,
    /// -Z (because of Bevy right hand rule.)
    Front,
    /// +Z (because of Bevy right hand rule.)
    Back,
}

/// Contains each block face a block can have in the order their `index` method returns.
pub const ALL_BLOCK_FACES: [BlockFace; 6] = [
    BlockFace::Right,
    BlockFace::Left,
    BlockFace::Top,
    BlockFace::Bottom,
    BlockFace::Front,
    BlockFace::Back,
];

impl BlockFace {
    /// Returns the index for each block face [0, 5].
    ///
    /// Useful for storing faces in an array.
    /// This index does not directly correspond to any `Direction` index. Use `direction` to convert.
    pub const fn index(&self) -> usize {
        match *self {
            Self::Right => 0,
            Self::Left => 1,
            Self::Top => 2,
            Self::Bottom => 3,
            Self::Front => 4,
            Self::Back => 5,
        }
    }

    /// Gets this block face from its index.
    ///
    /// This will panic if index is not between 0 and 5 inclusive.
    /// This index does not directly correspond to any `Direction` index. Use `direction` to convert.
    #[inline]
    pub fn from_index(index: usize) -> Self {
        match index {
            0 => Self::Right,
            1 => Self::Left,
            2 => Self::Top,
            3 => Self::Bottom,
            4 => Self::Front,
            5 => Self::Back,
            _ => panic!("BlockFace index {index} is not between 0 and 5 inclusive."),
        }
    }

    /// Returns the string representation of this face.
    pub const fn as_str(&self) -> &'static str {
        match *self {
            Self::Right => "right",
            Self::Left => "left",
            Self::Top => "top",
            Self::Bottom => "bottom",
            Self::Front => "front",
            Self::Back => "back",
        }
    }

    /// Gets the opposite face for this block face (example: [`BlockFace::Left`] -> [`BlockFace::Right`])
    pub fn inverse(&self) -> Self {
        match self {
            Self::Right => Self::Left,
            Self::Left => Self::Right,
            Self::Top => Self::Bottom,
            Self::Bottom => Self::Top,
            Self::Front => Self::Back,
            Self::Back => Self::Front,
        }
    }

    /// Returns the [`Direction`] this [`BlockFace`] points if the block and it's structure are not rotated.
    ///
    /// Most blocks have some rotation, so be careful to call the proper `BlockRotation` method instead if the block is rotated.
    pub fn direction(self) -> BlockDirection {
        match self {
            Self::Right => BlockDirection::PosX,
            Self::Left => BlockDirection::NegX,
            Self::Top => BlockDirection::PosY,
            Self::Bottom => BlockDirection::NegY,
            Self::Front => BlockDirection::NegZ, // IMPORTANT: Due to Bevy's right hand rule, "front" points negative Z.
            Self::Back => BlockDirection::PosZ,  // IMPORTANT: Due to Bevy's right hand rule, "back" points positive Z.
        }
    }
}

impl Display for BlockFace {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())?;

        Ok(())
    }
}
